keyring.go

```diff
@@ -193,34 +193,57 @@ func (k KeyRing) VerifyJSONs(ctx context.Context, requests []VerifyJSONRequest)
 		return nil, err
 	}
 
-	// TODO: we should distinguish here between expired keys, and those we don't have.
-	// If the key has expired, it's no use re-requesting it.
-	now := AsTimestamp(time.Now())
+	// Default to not hitting federation.
+	doFederationHit := false
+
 	keysFetched := map[PublicKeyLookupRequest]PublicKeyLookupResult{}
+	now := AsTimestamp(time.Now())
 	for req, res := range keysFromDatabase {
+		if res.ExpiredTS != PublicKeyNotExpired {
+			// The key is expired - it's not going to change so just return
+			// it and don't bother requesting it again.
+			keysFetched[req] = res
+			delete(keyRequests, req)
+			continue
+		}
 		if now > res.ValidUntilTS && res.ExpiredTS == PublicKeyNotExpired {
-			// We have passed the key validity period so we should re-request from
-			// the remote server (or the perspectives).
+			// We're past the validity for this key so we should really
+			// hit the fetchers regardless to renew it.
+			doFederationHit = true
 			continue
 		}
+		// The key isn't expired and it's inside the validity period so
+		// include it here.
 		keysFetched[req] = res
 	}
 
-	k.checkUsingKeys(requests, results, keyIDs, keysFetched)
-
-	// If we can verify using the keys from the database, don't make a federation call.
-	doFederationHit := false
-	for _, r := range results {
-		if r.Error != nil {
-			doFederationHit = true
-			break
+	if len(keysFetched) == len(keyRequests) {
+		// If our key requests are all satisfied then we can try performing
+		// a verification using our keys.
+		k.checkUsingKeys(requests, results, keyIDs, keysFetched)
+
+		// If we run into any errors when verifying using the keys that we
+		// have then we can hit federation and check for updated keys.
+		for _, r := range results {
+			if r.Error != nil {
+				doFederationHit = true
```

Comment:
I would combine this with `if !doFederationHit { ... }` so in other words what you're saying is:
- If `len(keysFetched) == len(keyRequests)` and there's no errors from `checkUsingKeys` then return results
- Else do federation hits.

Basically we don't need `doFederationHit`.

Reply:
The logic here is:

1. Ask for keys that we know about from the database:
    1. If *any of the keys* from the database are past their validity then we need to hit federation anyway in order to get the renewed validities.
2. If we successfully satisfy all requested keys, try to `checkWithKeys`:
    1. If it succeeds (i.e. we haven't claimed we need to hit federation), return the results and stop here.
    2. If it doesn't succeed, hit federation in case one of the keys isn't what we thought it was.
3. If we didn't successfully satisfy all requested keys, then hit federation to get the keys we don't know about yet.

I'm not sure removing `doFederationHit` makes that pattern any clearer.

Reply:
I think it does maker it clearer overall. Then you can just do:
```go
if len(keysFetched) == len(keyRequests) {
    k.checkUsingKeys(requests, results, keyIDs, keysFetched)
    noErrors := true
    for _, r := range results {
        if r.Error != nil {
            noErrors := false
            break
        }
    }
    if noErrors {
        return results, nil
    }
}

// hit federation
```

Reply:
Have done that as a part of some greater magnifying-glass-ing, and I've also made one or two more tweaks. As far as I can tell from copious logging, this *should* be doing the right thing now.